Feature/logger

ssysoenko opened the pull request:
Добавлен логгер из CVSLogger. 
Рефакторинг логгера.
Фабрики возвращают результат в `CVSOutcome`.

shkryab committed:
Refactoring

Change factory result to CVSOutcome.
Copy logger from CVSLogger.
Save the config for each logger.
Use seved config in argumentPreprocessor.
Headeronly opencvHelper.
Fix loggers tests.

shkryab committed:
Fix install

shkryab committed:
Fix export

Fix format (need update clang-format)

shkryab committed:
Add componet var to install helper

Add component for opencvHelper
Set deps for  opencvHelper package

shkryab committed:
Fix tests

shkryab committed:
Remove opencvHelper.cpp

shkryab committed:
Update version

shkryab committed:
Fix

Fix cmake helpers
Fix install targets
Fix dependencies

shkryab committed:
Add helpers

Add typedef for boost::property_tree::ptree
Add throwWithNested function

efimov-evgeniy commented on logger/include/cvs/logger/opencvHelper.hpp:
```diff
@@ -0,0 +1,73 @@
+      std::filesystem::path save_path = config.img_path;
+      if (config.img_subfolder)
+        save_path /= Config::subfolderName();
+      save_path = save_path / logger->name() / spdlog::level::to_short_c_str(lvl);
```
А почему тут тоже не использовать `/=`?

ssysoenko replied:
Потому, что `logger->name() / spdlog::level::to_short_c_str(lvl);` не работатет.

efimov-evgeniy commented on logger/src/logging.cpp:
```diff
@@ -0,0 +1,152 @@
+    auto loggers_list = config.get_child_optional("loggers");
+    if (loggers_list) {
+      for (auto l : *loggers_list) {
+        auto cfg = LoggerConfig::make(l.second);
```
Переменная `cfg` не используется?

efimov-evgeniy commented on logger/src/logging.cpp:
```diff
@@ -0,0 +1,152 @@
+
+        configureLogger(**logger_opt, *cfg_struct_opt);
+
+        configurations(cfg_struct_opt->name) = cfg;
```
Может, как-то переименовать функцию? Разделить на константную `at` и записывающую `set`? Имхо присвоение вызову функции - не самая очевидная механика.

ssysoenko replied:
В функции `static` переменная. С глобальными и т.п. переменными были какие-то проблемы (не сейчас, а раньше и я уже забыл что конкретно и пишу по инерции).

efimov-evgeniy replied:
Не понял про глобальные переменные. Я о функции исключительно. Сделать что-то вроде `static const Configuration& at(std::string name)` и `static void/bool set(std::string name, Configuration new_configuration)`

ssysoenko replied:
Идея сделать глобальную мапу с настройками логера. Нужно чтобы можно было сохранять дерево настроек для конкретного логера и потом использовать их в `argumentPreprocessor`. Это перерасход ресурсов, но нет ограничения, что необходимо для каждой специализации `argumentPreprocessor` дергать свой `initLogger` (в `initLogger` парсятся и инициализируются базовые параметры логера, а каждый препроцессор, если надо, парсит свои сам в момент вызова).
Можно сделать глобальную переменную и получить проблемы с несколькими экземплярами в момент линковки, или использовать функцию, в которой создается статическаая еперменная и возвращать ссылку на эту переменную (этот подход, на сколько я помню, лишен проблем с несколькими экземплярами). Именно потому как должна быть такая функция - невозможно сделать `at` и `set` (ну или такие функции сделать, но `configurations` никуда не денется).

efimov-evgeniy commented on logger/src/tools/fpsLogger.cpp:
```diff
@@ -0,0 +1,165 @@
+void FpsLogger::Private::update(duration last_dur) {
+  auto f = fps(last_dur);
+  if (use_lock)
+    update_mutex.lock();
```
Если сделать `std::optional<std::shared_lock>` (к примеру), можно будет проверять один раз и использовать RAII. И вместо `update_mutex` + `use_lock` можно использовать `std::optional<std::shared_mutex>`.

ssysoenko replied:
Мьюткс копировать нельзя.

efimov-evgeniy replied:
Так в моём предложении и нет копирования, вроде бы.

ssysoenko replied:
`std::optional` пытается определитиь `operator=(const std::optional&)` и выдает ошибку.

efimov-evgeniy replied:
Это `std::optional<std::shared_mutex>`? У меня просто так сделано в `modules/ROSModule/Mediator`, и там всё работает.

Aldrog replied:
https://en.cppreference.com/w/cpp/utility/optional/operator%3D

> Overload (2) is defined as deleted unless std::is_copy_constructible_v<T> and std::is_copy_assignable_v<T> are both true.

Может быть, ты его где-то таки пытаешься скопировать и ошибка на это выдаётся?

efimov-evgeniy replied:
Вероятно, ты где-то пытаешься вызвать это оператор присваивания.

ssysoenko replied:
Пытался делать и так и так:
```
std::optional<std::shared_mutex> opt = std::shared_mutex();
std::optional<std::shared_mutex> opt = std::make_optional(std::shared_mutex());
```

efimov-evgeniy replied:
`std::optional<std::shared_mutex> opt(std::make_optional<std::shared_mutex>());`

Aldrog replied:
Можно так:
```
std::optional<std::shared_mutex> opt{std::in_place};
```
Вот перезаписать этот `optional` не получится, т.к. мутекс не только не копируется, но даже не перемещается.

ssysoenko replied:
Короче, не получается.

efimov-evgeniy replied:
```  
  std::optional<std::shared_mutex> opt(std::in_place);
  std::optional<std::scoped_lock<std::shared_mutex>> lock;
  if (opt) {
    lock.emplace(*opt);
  }

Aldrog replied:
@efimov-evgeniy `setUseLock` ты не реализуешь так.

efimov-evgeniy replied:
Кстати, да: а зачем это реализовывать? Предполагается, что логгер может переходить от синхронизации и обратно в процессе работы?
RAII лок в любом случае можно реализовать.

efimov-evgeniy replied:
В текущем виде это вообще дедлок потенциальный

shkryab committed:
Fix unused variable

shkryab committed:
FpsLogger refactoring

efimov-evgeniy commented on logger/src/tools/fpsLogger.cpp:
```diff
@@ -36,130 +78,191 @@ class FpsLogger::Private {
+namespace cvs::logger::tools {
+
+template <bool ThreadSafe>
+class FpsLogger<ThreadSafe>::Private {
```
А что это за конструкция?

ssysoenko replied:
pimpl

efimov-evgeniy commented on logger/src/tools/fpsLogger.cpp:
```diff
@@ -36,130 +78,191 @@ class FpsLogger::Private {
-    m->started = true;
+template <bool ThreadSafe>
+void FpsLogger<ThreadSafe>::start() {
+  if (!started_) {
```
Это, вообще говоря, небезопасно. Следует использовать атомарный бул и `compare_exchange...`

efimov-evgeniy commented on logger/src/tools/fpsLogger.cpp:
```diff
@@ -0,0 +1,268 @@
+  if (!started_)
+    return;
+
+  if (total_cnt) {
```
А почему тут нет никакой синхронизации?

efimov-evgeniy commented on logger/src/tools/fpsLogger.cpp:
```diff
@@ -0,0 +1,268 @@
+
+  time_point prev_frame_point;
+
+  duration report_period = 10min;
```
Стоит, наверное, обозначит как `const`, иначе вопросы к синхронизации

efimov-evgeniy commented on logger/src/tools/fpsLogger.cpp:
```diff
@@ -0,0 +1,268 @@
+
+template <bool ThreadSafe>
+double FpsLogger<ThreadSafe>::fps() const {
+  if (total_dur.count())
```
Это тоже небезопасно. Между проверкой (которая, кстати говоря, тоже вряд ли атомарна) и вызовом `fps` может вклиниться `clear`, и получится деление на 0.

shkryab committed:
FPSLogger refactoring

FPSCounter insted of FPSLogger
Logger config translators refactoring
Throw exception helper

efimov-evgeniy commented on logger/src/tools/fpsCounter.cpp:
```diff
@@ -0,0 +1,135 @@
+  void start() {
+    bool expected = false;
+    if (started.compare_exchange_strong(expected, true)) {
+      last_frame_time = clock::now();
```
Опасно. `newFrame` может вызваться после проверки, и `last_frame_time` не успеет установиться

efimov-evgeniy replied:
Я бы вообще предложил остановиться на использовании исключительно 'shared_mutex' для синхронизации и не мучиться. Если хочется производительности, то тут логично было бы использовать аналог RCU.

efimov-evgeniy commented on logger/src/tools/fpsCounter.cpp:
```diff
@@ -0,0 +1,135 @@
+    auto total_duration_copy = total_duration;
+    auto smma_fps_copy       = smma_fps;
+
+    lock.unlock();
```
Лучше сделать скоуп ( `{}` ), чем вручную анлочить

ssysoenko replied:
Чем лучше?

efimov-evgeniy replied:
Лучше использовать механизмы RAII - они более очевидны. Будет видно, что у тебя отдельный блок с синхронизацией.

shkryab committed:
fix include

shkryab committed:
Fix start/stop

shkryab committed:
PR fix

ssysoenko committed:
Add namespace for version

ssysoenko committed:
Add duration param

ssysoenko committed:
FPSCounter for refactoring

ssysoenko committed:
Fix FPSCounter

Fix fps
Add latency counter

efimov-evgeniy commented on logger/src/tools/fpsCounter.cpp:
```diff
@@ -48,12 +53,14 @@ class FPSCounter : public IFPSCounter {
       total_duration += last_frame_duration;
+
       ++frame_count;
+      start_points.pop();
```
Так и предполагается, что сначала читается конец очереди, а потом удаляется первый элемент?

shkryab committed:
Fix FPSCounter

Fix ltency queue
Fix clear()

roman-orekhov commented on logger/include/cvs/logger/loggable.hpp:
```diff
@@ -0,0 +1,26 @@
+class Loggable {
+ public:
+  explicit Loggable(const common::Properties& cfg) { log = *createLogger(cfg); }
+  explicit Loggable(const std::string& name = boost::core::demangle(typeid(T).name())) { log = *createLogger(name); }
```
Зачем весь Loggable делать шаблоном, если шаблонного у него -- один конструктор, который и можно сделать шаблонным?

ssysoenko replied:
Тогда придется каждый раз вызывать конструктор чтобы получить имя логера. А так можно воспользоваться дефортным именем.

roman-orekhov replied:
Не понял вообще. Если есть объект уже, то у него геттер на имя должен быть. Если нет, то зачем тебе имя логгера, когда он не создан ещё? Можешь кодом описать неудобный в предлагаемом варианте use case?

ssysoenko replied:
Возможно я совсем не понял предлагаемый use case. Шаблон сделан чтобы можно было избежать определения конструктора только для того чтобы задать имя логера, особенно когда имя логера вообще не важно, а важно чтоб он просто был. Я не понимаю откуда возьмется имя логера если убрать шаблон? 

От `Loggable` наследуется всякий кто хочет получить именованный логгер. Шаблон нужен только для того, чтобы задать логеру вменяемое имя "автоматически". Если убрать шаблон - придется всякий раз вызывать конструктор руками и передавать в него имя (ну или тип и из него получиать имя). Шаблоный конструктор тоже придется вызывать руками т.к. компилятор для него тип автоматически не выведет.

Думаю нужно добавить вариант реализации, где в шаблон передается сразу имя логера (потипу того как в конфиге коментарии полей задаются) чтобы можно было делать 2мя способами:
1. `class Q : public Loggable<Q>` - тут получаем имя через `boost::core::demangle(typeid(Q).name())`
2. `class Q : public Loggable<CVS_CONSTEXPRSTRING("Qwerty")>` - тут напрямую буквами задаем имя логера

roman-orekhov replied:
Понял, о чём речь. Я просто хотел иметь шаблон поменьше. Тогда я бы сделал template<class T> class Loggable : public LoggableImpl с шаблонным конструктором и using второго конструктора, а вся нешаблонная часть легла бы в LoggableImpl

roman-orekhov commented on logger/include/cvs/logger/tools/fpsCounter.hpp:
```diff
@@ -0,0 +1,46 @@
+class IFPSCounter {
+ public:
+  using CounterClock     = std::chrono::system_clock;
+  using CounterDuration  = CounterClock::duration;
```
Зачем в типах префикс Counter? К ним всё равно доступ через IFPSCounter:: будет. Если в типе префикс другого типа, это C-style, и в целом же решается через :: в C++...

ssysoenko replied:
Это не задумка, просто так получилось в процессе.

roman-orekhov commented on logger/include/cvs/logger/tools/fpsCounter.hpp:
```diff
@@ -0,0 +1,46 @@
+
+  virtual void clear() = 0;
+
+  virtual double statistics() const = 0;
```
После рефакторинга пропала та единственная фича, ради которой я и затевал частичную thread-safety изначально, -- возможность забрать текущую статистику по запросу "со стороны" через вызов метода.

ssysoenko replied:
Да. Это не досмотрели
